### PR TITLE
Limit label selector to Ingress factory.

### PR DESCRIPTION
**Pull request by timoreimann**
### What does this PR do?

Limit the Kubernetes label selector to the Ingress factory.

### Motivation

Fix a regression when we switched to the newer shared informer factory.

### Reviewer Notes

@yue9944882 could you please take a look and let me know if this seems good to you?

I'm also open to any suggestions how we could possibly implement this more elegantly.

Thanks!

Refs #3128

### CODE REVIEW

**Comment by timoreimann** [comment]
@ldez done.

**Comment by yue9944882** [comment]
I didn't really notice that this labelSelector should only be applied for ingress. How about renaming the parameter to `ingressLabelSelector` to emphasise it?

File: provider/kubernetes/client.go [hunk 1]
```diff
@@ -160,7 +171,7 @@ func (c *clientImpl) WatchAll(namespaces Namespaces, labelSelector string, stopC
 // GetIngresses returns all Ingresses for observed namespaces in the cluster.
 func (c *clientImpl) GetIngresses() []*extensionsv1beta1.Ingress {
 	var result []*extensionsv1beta1.Ingress
-	for ns, factory := range c.factories {
+	for ns, factory := range c.ingressFactories {
 		ings, err := factory.Extensions().V1beta1().Ingresses().Lister().List(labels.Everything())
```
[the remark this answers is not in the document]
Comment (timoreimann) [reply provider/kubernetes/client.go]: Ah yeah, that'd be much better (and pretty much what I was longing for but failed to see).

Let me update the PR...

**Comment by dtomcej** [comment]
Code looks good, but do you think we should have a test for this? since we merged a regression without it being caught?

File: provider/kubernetes/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (yue9944882) [reply provider/kubernetes/client.go]: @timoreimann We can also use `cache.FilteringResourceEventHandler` to specify a filter func to perform label selector matching for ingress resources. See examples here :)
[https://github.com/kubernetes/kubernetes/blob/dfd6f581b5c8830c760d2778b299c3fd5e7aa242/pkg/scheduler/factory/factory.go#L186](https://github.com/kubernetes/kubernetes/blob/dfd6f581b5c8830c760d2778b299c3fd5e7aa242/pkg/scheduler/factory/factory.go#L186)

**Comment by timoreimann** [comment]
@dtomcej 

>Code looks good, but do you think we should have a test for this? since we merged a regression without it being caught?

I agree with you that we _should_ have a test (in fact, tests for most of our low-level Kubernetes integration functionality). However, I don't think that unit tests would be really valuable in this case: The code in `client.go` largely depends on the client-go library, which means that any tests we write would heavily depend on the fake implementation that comes with it. IMHO, it'd be the classical case of testing against a test double without really testing the production code.

What I think we need are proper integration tests for Kubernetes. I've been working on #1889 for quite a while; it's not quite done, but I hope to finish it rather soon. Personally, I'd rather invest into these when it comes to validating our integration logic.

File: provider/kubernetes/client.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (timoreimann) [reply provider/kubernetes/client.go]: Updated the PR.

PTAL again.

File: provider/kubernetes/client.go [hunk 2]
```diff
@@ -215,8 +209,18 @@ func (c *clientImpl) lookupNamespace(ns string) string {
 	return ns
 }
 
-func newResourceEventHandler(events chan<- interface{}) cache.ResourceEventHandler {
-	return &resourceEventHandler{events}
+func (c *clientImpl) newResourceEventHandler(events chan<- interface{}) cache.ResourceEventHandler {
+	return &cache.FilteringResourceEventHandler{
+		FilterFunc: func(obj interface{}) bool {
+			// Ignore Ingresses that do not match our custom label selector.
+			if ing, ok := obj.(*extensionsv1beta1.Ingress); ok {
```
Comment (yue9944882) [inline provider/kubernetes/client.go]: LGTM now 💯

**Comment by timoreimann** [comment]
I manually tested the PR both with and without a configured label selector. Looks like it's working as expected.

File: provider/kubernetes/kubernetes.go [hunk 3]
```diff
@@ -45,24 +46,37 @@ type Provider struct {
...
+	LabelSelector          string     `description:"Kubernetes Ingress label selector to use" export:"true"`
 	IngressClass           string     `description:"Value of kubernetes.io/ingress.class annotation to watch for" export:"true"`
 	lastConfiguration      safe.Safe
 }
 
-func (p *Provider) newK8sClient() (Client, error) {
+func (p *Provider) newK8sClient(ingressLabelSelector string) (Client, error) {
+	ingLabelSel, err := labels.Parse(ingressLabelSelector)
+	if err != nil {
+		return nil, fmt.Errorf("invalid ingress label selector: %s", ingressLabelSelector)
```
Comment (nmengin) [inline provider/kubernetes/kubernetes.go]: WDYT about using `%q` instead of `%s` ?

File: provider/kubernetes/kubernetes.go [hunk 4]
```diff
@@ -45,24 +46,37 @@ type Provider struct {
...
+	}
+	log.Infof("ingress label selector is: %s", ingLabelSel)
```
Comment (nmengin) [inline provider/kubernetes/kubernetes.go]: WDYT about using `%q` instead of `%s` ?

**Pull request merged**